<a href="https://colab.research.google.com/github/josedanielisidororeyes/Advanced_Data_Engineering/blob/main/Operaciones_b%C3%A1sicas_con_DataFrames_en_PySpark_(Titanic).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Nombre del Alumno: José Daniel Isidoro Reyes
#Matrícula: 261552
#Matería: Ingeniería de Datos Avanzada
#Nombre de la Tarea: Operaciones básicas con DataFrames en PySpark (Titanic)
#Fecha: 29/03/2026

In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, count, when, isnan

# Creación de sesión en Spark
spark = SparkSession.builder \
    .appName('Tarea Titanic: Operaciones Básicas') \
    .master('local[*]') \
    .getOrCreate()


Se configura la  sesión para poder interatuar con dataframes en Spark.

#1) Carga del Dataset

In [ ]:
df =  spark.read.csv('/content/drive/MyDrive/Ingeniería de Datos Avanzada/Titanic-Dataset.csv', header =  True, inferSchema  =  True)


#2) Exploración Inicial

In [ ]:
# Visualización de las primeras 5 filas
df.show(5)

+-----------+--------+------+--------------------+------+----+-----+-----+----------------+-------+-----+--------+
|PassengerId|Survived|Pclass|                Name|   Sex| Age|SibSp|Parch|          Ticket|   Fare|Cabin|Embarked|
+-----------+--------+------+--------------------+------+----+-----+-----+----------------+-------+-----+--------+
|          1|       0|     3|Braund, Mr. Owen ...|  male|22.0|    1|    0|       A/5 21171|   7.25| NULL|       S|
|          2|       1|     1|Cumings, Mrs. Joh...|female|38.0|    1|    0|        PC 17599|71.2833|  C85|       C|
|          3|       1|     3|Heikkinen, Miss. ...|female|26.0|    0|    0|STON/O2. 3101282|  7.925| NULL|       S|
|          4|       1|     1|Futrelle, Mrs. Ja...|female|35.0|    1|    0|          113803|   53.1| C123|       S|
|          5|       0|     3|Allen, Mr. Willia...|  male|35.0|    0|    0|          373450|   8.05| NULL|       S|
+-----------+--------+------+--------------------+------+----+-----+-----+------

In [ ]:
# Visualización del tipo de esquema el dataframe
df.printSchema()

root
 |-- PassengerId: integer (nullable = true)
 |-- Survived: integer (nullable = true)
 |-- Pclass: integer (nullable = true)
 |-- Name: string (nullable = true)
 |-- Sex: string (nullable = true)
 |-- Age: double (nullable = true)
 |-- SibSp: integer (nullable = true)
 |-- Parch: integer (nullable = true)
 |-- Ticket: string (nullable = true)
 |-- Fare: double (nullable = true)
 |-- Cabin: string (nullable = true)
 |-- Embarked: string (nullable = true)



<p align = "Justify"

Se observa que el esquema permite valores nulos para todas las columnas. De igual modo, se observa que variables categoricas como Embarked, Cabin, Sex están categorizadas como texto, lo cual requiere transformación para poder ser utilizadas en modelos de Machine Learning.

# 3) Selección  de columnas

In [ ]:
df_seleccion =  df.select("Name", "Age", "Sex")
df_seleccion.show(5)

+--------------------+----+------+
|                Name| Age|   Sex|
+--------------------+----+------+
|Braund, Mr. Owen ...|22.0|  male|
|Cumings, Mrs. Joh...|38.0|female|
|Heikkinen, Miss. ...|26.0|female|
|Futrelle, Mrs. Ja...|35.0|female|
|Allen, Mr. Willia...|35.0|  male|
+--------------------+----+------+
only showing top 5 rows


Se realiza una selección de columnas para trabajar con un subconjunto de los datos.

#4) Filtrado de Datos

In [ ]:
df_filtered =  df.filter((col("Age") > 18) & (col("Sex") == "female"))
df_filtered.show(5)

+-----------+--------+------+--------------------+------+----+-----+-----+----------------+-------+-----+--------+
|PassengerId|Survived|Pclass|                Name|   Sex| Age|SibSp|Parch|          Ticket|   Fare|Cabin|Embarked|
+-----------+--------+------+--------------------+------+----+-----+-----+----------------+-------+-----+--------+
|          2|       1|     1|Cumings, Mrs. Joh...|female|38.0|    1|    0|        PC 17599|71.2833|  C85|       C|
|          3|       1|     3|Heikkinen, Miss. ...|female|26.0|    0|    0|STON/O2. 3101282|  7.925| NULL|       S|
|          4|       1|     1|Futrelle, Mrs. Ja...|female|35.0|    1|    0|          113803|   53.1| C123|       S|
|          9|       1|     3|Johnson, Mrs. Osc...|female|27.0|    0|    2|          347742|11.1333| NULL|       S|
|         12|       1|     1|Bonnell, Miss. El...|female|58.0|    0|    0|          113783|  26.55| C103|       S|
+-----------+--------+------+--------------------+------+----+-----+-----+------

Se realiza un filtrado para obtener a las mujeres mayores de 18 años.

#5) Valores Únicos

In [ ]:
print("Valores Únicos de la Columna Pclass ")
df.select("Pclass").distinct().show()

Valores Únicos de la Columna Pclass 
+------+
|Pclass|
+------+
|     1|
|     3|
|     2|
+------+



#6)Agrupación

In [ ]:
print("Pasajeros por Clase:")
df.groupBy("Pclass").count().show()

Pasajeros por Clase:
+------+-----+
|Pclass|count|
+------+-----+
|     1|  216|
|     3|  491|
|     2|  184|
+------+-----+



In [ ]:
print("Pasajeros por Sexo:")
df.groupby("Sex").count().show()

Pasajeros por Sexo:
+------+-----+
|   Sex|count|
+------+-----+
|female|  314|
|  male|  577|
+------+-----+



Se observa una mayor cantidad de pasajeros en la clase 3 con 491 registros contra 216 de la clase 1 y 184 de la clase 2. De igual modo, se observa una mayor cantidad de hombres en la muestra de datos.

# 7) Estadística Básica

In [ ]:
df.describe().show()

+-------+-----------------+-------------------+------------------+--------------------+------+------------------+------------------+-------------------+------------------+-----------------+-----+--------+
|summary|      PassengerId|           Survived|            Pclass|                Name|   Sex|               Age|             SibSp|              Parch|            Ticket|             Fare|Cabin|Embarked|
+-------+-----------------+-------------------+------------------+--------------------+------+------------------+------------------+-------------------+------------------+-----------------+-----+--------+
|  count|              891|                891|               891|                 891|   891|               714|               891|                891|               891|              891|  204|     889|
|   mean|            446.0| 0.3838383838383838| 2.308641975308642|                NULL|  NULL| 29.69911764705882|0.5230078563411896|0.38159371492704824|260318.54916792738| 32.20420

Se muestra  un analisis de estadistica descripva del conjunto de datos, por ejemplo, el 38.38% sobrevivió, la edad promedio es 29.69, ticket promedio de 260138.54, etc.

# 8 ) Valores Nulos

In [ ]:
df.select([
    count(when(isnan(c) | col(c).isNull(), c)).alias(c)
    if t in ("double", "float")
    else count(when(col(c).isNull(), c)).alias(c)
    for c, t in df.dtypes
]).show()

+-----------+--------+------+----+---+---+-----+-----+------+----+-----+--------+
|PassengerId|Survived|Pclass|Name|Sex|Age|SibSp|Parch|Ticket|Fare|Cabin|Embarked|
+-----------+--------+------+----+---+---+-----+-----+------+----+-----+--------+
|          0|       0|     0|   0|  0|177|    0|    0|     0|   0|  687|       2|
+-----------+--------+------+----+---+---+-----+-----+------+----+-----+--------+



Se observa 177 registros sin edad y 687 sin registros en la columna Cabin, los cuales podrían ser remplazados por el valor medio o la moda.